chain/chain/src/block_processing_utils.rs

```diff
@@ -0,0 +1,200 @@
+        Vec<Result<ApplyChunkResult, near_chain_primitives::Error>>,
+    )> {
+        if self.0.read().expect(POISONED_LOCK_ERR).applied_blocks.is_empty() {
+            return vec![];
+        }
+
+        let mut pool = self.0.write().expect(POISONED_LOCK_ERR);
+
+        let mut res = vec![];
+        let applied_blocks = pool.applied_blocks.drain().collect::<Vec<_>>();
+        for (block_hash, apply_result) in applied_blocks {
+            let (block, preprocess_result) = pool.preprocessed_blocks.remove(&block_hash).expect(
+                &format!("block {:?} in applied_blocks but not in preprocessed_blocks", block_hash),
+            );
+            res.push((block, preprocess_result, apply_result));
+        }
+
+        res
+    }
+
+    /// Returns whether the block has been added as preprocessed
+    pub(crate) fn is_block_preprocessed(&self, block_hash: &CryptoHash) -> bool {
+        self.0.read().expect(POISONED_LOCK_ERR).preprocessed_blocks.contains_key(block_hash)
+    }
+
+    /// Returns whether the pool is full
+    pub(crate) fn is_full(&self) -> bool {
+        self.0.read().expect(POISONED_LOCK_ERR).preprocessed_blocks.len() >= MAX_PROCESSING_BLOCKS
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::block_processing_utils::{BlockLike, BlockPreprocessInfo, BlockProcessingPool};
+    use actix::clock::{sleep, timeout};
+    use near_primitives::hash::{hash, CryptoHash};
+    use near_primitives::time::Duration;
+    use std::collections::HashSet;
+
+    #[derive(Clone)]
+    struct MockBlock {
+        hash: CryptoHash,
+    }
+    impl BlockLike for MockBlock {
+        fn hash(&self) -> &CryptoHash {
+            &self.hash
+        }
+    }
+
+    #[test]
+    // This test creates a scenario similar to how BlockProcessingPool will be used in chain.
+    // One thread adds preprocessed blocks to the pool and schedule other threads to add applied
+    // blocks to the pool. Then it waits and keeps checking if there are applied blocks in the pool
+    fn test_block_processing_pool_basic() {
+        actix::System::new().block_on(async {
+            timeout(Duration::from_secs(5), async {
+                let pool = BlockProcessingPool::new();
+                let mut hashes = HashSet::new();
+                for i in 0..10 {
+                    let hash = hash(&[i]);
+                    let res = pool.add_preprocessed_block(
+                        MockBlock { hash: hash.clone() },
+                        BlockPreprocessInfo::default(),
+                    );
+                    if res.is_ok() {
+                        hashes.insert(hash);
+                        let pool1 = pool.clone();
+                        let hash1 = hash.clone();
+                        // spwan another thread to add applied blocks later
+                        actix::spawn(async move {
+                            assert!(pool1.is_block_preprocessed(&hash1));
+                            sleep(Duration::from_secs(1)).await;
```

Comment:
I feel stronogly that sleeps in tests are a very bad practice which we should avoid. The badness comes from a combination of two factors: 

* one sleeping test is not bad -- it takes just a second
* one thousands sleeping tests are horrible

This is a boiling frog kind of situation, so I want to push rather hard against this practice :-)  


When writing concurrent tests, the standard practice is to use barrier synchronization to tests specific interleavings. It's seems a bit harder than using sleeps, *but* it actually easier to write a non-flaky test that way. For example, the way the test is written currently, the `assert!(pool.is_full());` could in theory fail, if the system is super over-loaded and the worker threads manage to finish first, despite the sleeps. 

Taking everything together, here's how I'd write this test: 

```rust
    #[test]
    // This test creates a scenario similar to how BlockProcessingPool will be used in chain.
    // One thread adds preprocessed blocks to the pool and schedule other threads to add applied
    // blocks to the pool. Then it waits and keeps checking if there are applied blocks in the pool
    fn test_block_processing_pool_basic() {
        let mut threads = Vec::new();
        let barrier = Arc::new(std::sync::Barrier::new(MAX_PROCESSING_BLOCKS + 1));
        let pool = BlockProcessingPool::new();
        let mut hashes = HashSet::new();
        for i in 0..10 {
            let hash = hash(&[i]);
            let res = pool.add_preprocessed_block(
                MockBlock { hash: hash.clone() },
                BlockPreprocessInfo::default(),
            );

            if res.is_ok() {
                hashes.insert(hash);
                threads.push(std::thread::spawn({
                    let pool = pool.clone();
                    let hash = hash.clone();
                    let barrier = Arc::clone(&barrier);
                    move || {
                        barrier.wait();
                        assert!(pool.is_block_preprocessed(&hash));
                        pool.add_applied_block(hash, vec![]).unwrap();
                    }
                }));
            } else {
                assert!(!pool.is_block_preprocessed(&hash));
            }
        }

        assert!(pool.is_full());
        for hash in hashes.iter() {
            assert!(pool.is_block_preprocessed(hash));
        }
        barrier.wait();
        // check applied blocks in the pool
        while !hashes.is_empty() {
            for (block, _, _) in pool.take_applied_blocks() {
                assert!(hashes.remove(block.hash()));
            }
        }

        for t in threads {
            t.join().unwrap()
        }
    }
```